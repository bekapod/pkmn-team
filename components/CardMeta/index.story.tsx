import { storiesOf } from "@storybook/react";
import React from "react";
import CardMeta from ".";

storiesOf("CardMeta", module).add(
  "default",
  (): JSX.Element => (
    <CardMeta
      id="1"
      items={[
        { label: "Item 1", value: "Value 1" },
        { label: "Item 2", value: 2 }
      ]}
    />
  )
);
