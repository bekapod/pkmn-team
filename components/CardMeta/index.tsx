import React from "react";
import styled from "styled-components/macro";
import * as variables from "../../helpers/variables";
import { lineHeight } from "../../helpers/verticalRhythm";
import Label from "../Label";

interface Props {
  id: string;
  items?: {
    label: string;
    value: string | number;
  }[];
}

const CardMetaWrapper = styled.div`
  display: flex;
  margin: ${variables.spacing.md}px 0;
`;

const CardMetaItem = styled.div`
  position: relative;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  flex-basis: 0;
  align-items: center;
  text-align: center;

  &::after {
    content: "";
    position: absolute;
    right: calc(${variables.spacing.xs}px / 2 * -1);
    top: 50%;
    width: ${variables.spacing.xs}px;
    height: 33px;
    background-color: ${variables.colors.grayLight};
    transform: translateY(-50%);
    border-radius: ${variables.spacing.xs}px;
  }

  &:last-child::after {
    display: none;
  }
`;

const CardMetaValue = styled.span`
  color: ${variables.colors.grayDark};
  font-size: ${variables.fontSizes.md}px;
  font-weight: 900;
  line-height: ${lineHeight("md")};
`;

const CardMeta = ({ id, items = [] }: Props): JSX.Element => (
  <CardMetaWrapper data-testid={`card-meta-${id}`}>
    {items.map(
      ({ label, value }): JSX.Element => (
        <CardMetaItem key={`Team: ${id}, Label: ${label}`}>
          <Label>{label}</Label>
          <CardMetaValue>{value}</CardMetaValue>
        </CardMetaItem>
      )
    )}
  </CardMetaWrapper>
);

export default CardMeta;
