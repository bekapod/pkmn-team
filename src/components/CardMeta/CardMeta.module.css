.wrapper {
  display: flex;
  margin-top: var(--spacing-4);
  margin-bottom: var(--spacing-4);
}

.item {
  position: relative;
  display: flex;
  flex-direction: column;
  flex: 1 0 auto;
  align-items: center;
  text-align: center;

  &::after {
    content: '';
    position: absolute;
    right: calc(var(--spacing-1) / 2 * -1);
    top: 50%;
    width: var(--spacing-2);
    height: var(--spacing-5);
    background-color: var(--color-indigo-100);
    transform: translateY(-50%);
    border-radius: var(--border-radius-xl);
  }

  &:last-child::after {
    display: none;
  }
}

.label {
  margin-bottom: var(--spacing-1);
}

.value {
  color: var(--color-indigo-800);
  font-size: var(--font-size-lg);
  line-height: var(--line-height-lg);
  font-weight: 900;
}
