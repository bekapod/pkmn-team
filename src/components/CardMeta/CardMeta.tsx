import type { FunctionComponent } from 'react';
import classNames from 'classnames';
import { Label } from '../Label';

export type CardMetaProps = {
  items?: {
    label: string;
    value: string | number;
  }[];
};

export const CardMeta: FunctionComponent<CardMetaProps> = ({ items = [] }) => (
  <dl role="associationlist" className={classNames('flex', 'my-4')}>
    {items.map(({ label, value }, idx) => (
      <div
        className={classNames(
          'relative',
          'flex',
          'flex-col',
          'flex-grow',
          'flex-shrink-0',
          'items-center',
          'text-center'
        )}
        key={label}
      >
        <Label
          className={classNames('mb-1')}
          as="dt"
          role="associationlistitemkey"
          aria-label={label}
        >
          {label}
        </Label>
        <dd
          role="associationlistitemvalue"
          aria-label={value.toString()}
          className={classNames(
            'm-0',
            'text-indigo-800',
            'text-lg',
            'font-black'
          )}
        >
          {value}
        </dd>
        {idx !== items.length - 1 && (
          <span
            role="presentation"
            className={classNames(
              'block',
              'absolute',
              '-right-1',
              'top-1/2',
              'w-2',
              'h-5',
              'bg-indigo-100',
              '-translate-y-1/2',
              'transform-gpu',
              'rounded-xl'
            )}
          />
        )}
      </div>
    ))}
  </dl>
);
