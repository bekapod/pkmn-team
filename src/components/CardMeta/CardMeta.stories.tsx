import { Meta, Story } from '@storybook/react/types-6-0';
import { CardMeta, CardMetaProps } from './CardMeta';

export default {
  title: 'Components/Card Meta',
  component: CardMeta,
  args: {
    id: '1',
    items: [
      { label: 'Item 1', value: 'Value 1' },
      { label: 'Item 2', value: 2 }
    ]
  }
} as Meta<CardMetaProps>;

export const cardMeta: Story<CardMetaProps> = args => <CardMeta {...args} />;
