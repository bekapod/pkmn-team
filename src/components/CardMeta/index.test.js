// @flow
import React from "react";
import renderer from "react-test-renderer";
import CardMeta from ".";

describe("<CardMeta />", () => {
  it("renders without crashing", () => {
    const tree = renderer.create(
      <CardMeta
        id="1"
        items={[
          { label: "Item 1", value: "Value 1" },
          { label: "Item 2", value: 2 }
        ]}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("renders without crashing when no items are passed", () => {
    const tree = renderer.create(<CardMeta />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
