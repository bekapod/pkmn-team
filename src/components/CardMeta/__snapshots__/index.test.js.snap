// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardMeta /> renders without crashing 1`] = `
<div
  className="CardMeta__CardMetaWrapper-sc-6wqr52-0 hgGiwQ"
>
  <div
    className="CardMeta__CardMetaItem-sc-6wqr52-1 bDCDfU"
  >
    <span
      className="CardMeta__CardMetaLabel-sc-6wqr52-2 bZDTki"
    >
      Item 1
    </span>
    <span
      className="CardMeta__CardMetaValue-sc-6wqr52-3 dquxYs"
    >
      Value 1
    </span>
  </div>
  <div
    className="CardMeta__CardMetaItem-sc-6wqr52-1 bDCDfU"
  >
    <span
      className="CardMeta__CardMetaLabel-sc-6wqr52-2 bZDTki"
    >
      Item 2
    </span>
    <span
      className="CardMeta__CardMetaValue-sc-6wqr52-3 dquxYs"
    >
      2
    </span>
  </div>
</div>
`;

exports[`<CardMeta /> renders without crashing when no items are passed 1`] = `
<div
  className="CardMeta__CardMetaWrapper-sc-6wqr52-0 hgGiwQ"
/>
`;
